master/internal/resourcemanagers/resource_pool.go

```diff
@@ -344,12 +356,42 @@ func (rp *ResourcePool) Receive(ctx *actor.Context) error {
 func (rp *ResourcePool) receiveAgentMsg(ctx *actor.Context) error {
 	switch msg := ctx.Message().(type) {
 	case sproto.AddAgent:
-		ctx.Log().Infof("adding agent: %s", msg.Agent.Address().Local())
+		agentID := msg.Agent.Address().Local()
+		ctx.Log().Infof("adding agent: %+v", agentID)
 		rp.agents[msg.Agent] = true
+		err := rp.db.AddAgent(&model.AgentStats{
+			ResourcePool: rp.config.PoolName,
+			AgentID:      &agentID,
+			Slots:        msg.Slots,
+			StartTime:    time.Now().UTC().Truncate(time.Millisecond),
+		})
+		if err != nil {
+			ctx.Log().WithError(err)
+		}
 
 	case sproto.RemoveAgent:
-		ctx.Log().Infof("removing agent: %s", msg.Agent.Address().Local())
+		agentID := msg.Agent.Address().Local()
+		ctx.Log().Infof("removing agent: %s", agentID)
 		delete(rp.agents, msg.Agent)
+		endTime := time.Now().UTC().Truncate(time.Millisecond)
+		err := rp.updateAgentEndStats(agentID, &endTime)
```

Comment:
what happens if master crashes, and RemoveAgent for some agentID is never recorded?

Reply:
also I don't believe this supports user's agent disable/draining.

Reply:
just a note that we have to handle the first type of issue for allocations, too. https://github.com/determined-ai/determined/blob/1401d33f5b75ec23580a1ff4164bd9c7168e3d61/master/internal/db/postgres_tasks.go#L209-L226.